File: 2022/02/28/reverseVowels.ts
/**
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 * @author Andrew Jarombek
 * @since 2/28/2022
 */

function reverseVowels(s: string): string {
    let start = 0;
    let end = s.length - 1;
    let str = s.split("");
    let vowels = new Set(['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']);

    while (start < end) {
        while (start < end && !vowels.has(str[start])) {
            start++;
        }

        while (end > start && !vowels.has(str[end])) {
            end--;
        }

        if (start < end) {
            let temp = str[start];
            str[start] = str[end];
            str[end] = temp;
            start++;
            end--;
        }
    }

    return str.join("");
}

File: 2022/07/16/java/ReverseVowels.java
/**
 * https://leetcode.com/problems/reverse-vowels-of-a-string/
 * @author Andrew Jarombek
 * @since 7/16/2022
 */

import java.util.HashSet;
import java.util.Set;

public class ReverseVowels {
    public static String reverseVowels(String s) {
        Set<Character> vowels = new HashSet<>();
        vowels.add('a');
        vowels.add('e');
        vowels.add('i');
        vowels.add('o');
        vowels.add('u');
        vowels.add('A');
        vowels.add('E');
        vowels.add('I');
        vowels.add('O');
        vowels.add('U');

        char[] chars = s.toCharArray();
        int start = 0;
        int end = chars.length - 1;

        while (start < end) {
            while (start < end && !vowels.contains(chars[start])) {
                start++;
            }

            while (start < end && !vowels.contains(chars[end])) {
                end--;
            }

            char temp = chars[start];
            chars[start] = chars[end];
            chars[end] = temp;

            start++;
            end--;
        }

        return String.valueOf(chars);
    }

    public static void main(String[] args) {
        assert reverseVowels("hello").equals("holle");
    }
}
